File: 02_selection_sort/javascript/01_selection_sort.js
'use strict';
/**
 * Finds the index of the element with the smallest value in the array
 * @param {Array} array Source array
 * @returns {number} Index of the element with the smallest value
 */
const findSmallest = (arr) => {
  let [smallestElement] = arr;
  let smallestIndex = 0;
  for (let i = 1; i < arr.length; i++) {
    const el = arr[i];
    if (el >= smallestElement) continue;
    smallestElement = el;
    smallestIndex = i;
  }
  return smallestIndex;
};

/**
 * Sort array by increment
 * @param {Array} array Source array
 * @returns {Array} New sorted array
 */
const selectionSort = (arr) => {
  const size = arr.length;
  const result = new Array(size).fill(0);
  for (let i = 0; i < size; i++) {
    const smallestIndex = findSmallest(arr);
    const [smallestElement] = arr.splice(smallestIndex, 1);
    result[i] = smallestElement;
  }
  return result;
};

const sourceArray = [5, 3, 6, 2, 10];
const sortedArray = selectionSort(sourceArray);

console.log('Source array - ', sourceArray); // [5, 3, 6, 2, 10]
console.log('New sorted array - ', sourtedArray); // [2, 3, 5, 6, 10]

File: 02_selection_sort/R/01_selection_sort.R
# Finds the smallest value in an array
find_smallest <- function(arr) {
  # Stores the smallest value
  smallest = arr[1]
  # Stores the index of the smallest value
  smallest_index = 1
  for (i in 1:length(arr)) {
    if (arr[i] < smallest) {
      smallest_index = i
      smallest = arr[i]
    }
  }
  return(smallest_index)
}

# Sort array
selection_sort <- function(arr) {
  newArr = c()
  for (i in 1:length(arr)) {
    # Finds the smallest element in the array and adds it to the new array
    smallest = find_smallest(arr)
    newArr = c(newArr, arr[smallest])
    # Removes that smallest element from the original array
    arr = arr[-smallest]
  }
  return(newArr)
}

print(selection_sort(c(5, 3, 6, 2, 10)))
  